src/Raven.Server/Documents/Sharding/Comparers/DocumentsComparer.cs

```diff
@@ -10,31 +11,70 @@
 
 namespace Raven.Server.Documents.Sharding.Comparers;
 
-public class DocumentsComparer : IComparer<BlittableJsonReaderObject>
+public sealed class DocumentsComparer : IComparer<BlittableJsonReaderObject>
 {
     private readonly OrderByField[] _orderByFields;
     private readonly bool _extractFromData;
+    private readonly Random[] _randoms;
+    private readonly int[] _orderToRandom;
 
     public DocumentsComparer(OrderByField[] orderByFields, bool extractFromData)
     {
         _orderByFields = orderByFields;
         _extractFromData = extractFromData;
+        int randomCount = RandomComparersCount(_orderByFields);
```

Comment:
QueryMetadata already has `HasOrderByRandom`, you need to use it

Reply:
Done